File: frontend/src/api/events.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || '/api'

export interface EventRecord {
  id: number
  source_event_id?: number
  source: string
  source_label?: string
  source_category?: string
  event_type?: string
  signal_key?: string
  disposition?: string
  disposition_reason?: string
  decision_confidence?: number
  cluster_key?: string
  correlation_key?: string
  signal_family?: string
  external_event_id?: string
  dedup_key: string
  site_id?: number
  netbox_device_id?: number
  host?: string
  name: string
  severity: string
  severity_level: number
  status: string
  acknowledged: boolean
  occurred_at?: string
  resolved_at?: string
  last_seen_at?: string
  payload: Record<string, any>
  recommended_skill_code?: string
  case_id?: number
  case_code?: string
  created_at?: string
  updated_at?: string
}

export type EventItem = EventRecord

export interface EventListResponse {
  page: {
    total: number
    skip: number
    limit: number
    returned: number
    has_more: boolean
  }
  events: EventRecord[]
}

export interface EventClusterItem {
  cluster_key: string
  correlation_key: string
  title: string
  event_count: number
  source_categories: string[]
  dispositions: Record<string, number>
  case_count: number
  ticket_count: number
  highest_severity: string
  host?: string
  site_id?: number
  netbox_device_id?: number
  latest_occurred_at?: string
  signal_family?: string
  device_name?: string
  device_role?: string
  site_name?: string
  topology_hint?: string
  root_cause_candidate?: string
  adjacent_devices: string[]
  link_count: number
  impact_scope?: string
}

export interface RootCauseCandidateItem {
  candidate_key: string
  title: string
  root_cause_candidate: string
  site_name?: string
  signal_family?: string
  score: number
  ranking_reason: string
  merged_cluster_count: number
  event_count: number
  case_count: number
  ticket_count: number
  source_categories: string[]
  adjacent_devices: string[]
  representative_device?: string
  impact_scope?: string
  recommended_actions: Array<{
    priority_order: number
    action_type: string
    title: string
    reason: string
    mode: string
  }>
}

export interface EventRelationsResponse {
  event_id: number
  ticket: Record<string, any>
  linked_case?: {
    case_id: number
    case_code: string
    created_at?: string
  } | null
  linked_tasks: Array<{
    task_id: number
    task_code: string
    status: string
    source_model?: string
    case_id?: number | null
    recommended_skill_code?: string
    created_at?: string
  }>
}

export interface PlaybookDraftCheckResponse {
  success: boolean
  message: string
  event_id: number
  playbook_check: {
    passed?: boolean
    errors?: string[]
    warnings?: string[]
  }
  playbook_yaml?: string
}

export const eventsApi = {
  async listEvents(params?: {
    status?: string
    severity?: string
    source?: string
    event_type?: string
    source_category?: string
    disposition?: string
    site_id?: number
    netbox_device_id?: number
    skip?: number
    limit?: number
  }): Promise<EventListResponse> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events` : `${API_BASE_URL}/events`
    const response = await axios.get(apiUrl, { params })
    return response.data
  },

  async listClusters(params?: {
    status?: string
    severity?: string
    source?: string
    disposition?: string
    limit?: number
  }): Promise<{ clusters: EventClusterItem[] }> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/clusters` : `${API_BASE_URL}/events/clusters`
    const response = await axios.get(apiUrl, { params })
    return response.data
  },

  async listRootCauses(params?: {
    status?: string
    severity?: string
    source?: string
    disposition?: string
    limit?: number
  }): Promise<{ items: RootCauseCandidateItem[] }> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/root-causes` : `${API_BASE_URL}/events/root-causes`
    const response = await axios.get(apiUrl, { params })
    return response.data
  },

  async getMode(): Promise<{ message: string }> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/mode` : `${API_BASE_URL}/events/mode`
    const response = await axios.get(apiUrl)
    return response.data
  },

  async dispatchReadonly(
    eventId: number
  ): Promise<{ success: boolean; message: string; task_id?: number | null; case_id?: number | null; case_code?: string | null; playbook_check?: Record<string, any> }> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/${eventId}/dispatch-readonly` : `${API_BASE_URL}/events/${eventId}/dispatch-readonly`
    const response = await axios.post(apiUrl, {})
    return response.data
  },

  async createTicket(eventId: number, payload?: { title?: string; description?: string; priority?: string }): Promise<{ success: boolean; message: string; ticket_id?: string; provider?: string }> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/${eventId}/ticket` : `${API_BASE_URL}/events/${eventId}/ticket`
    const response = await axios.post(apiUrl, payload || {})
    return response.data
  },

  async getRelations(eventId: number): Promise<EventRelationsResponse> {
    const apiUrl = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/events/${eventId}/relations` : `${API_BASE_URL}/events/${eventId}/relations`
    const response = await axios.get(apiUrl)
    return response.data
  },

  async generatePlaybookDraft(eventId: number, includePlaybook: boolean = true): Promise<PlaybookDraftCheckResponse> {
    const apiUrl = API_BASE_URL.startsWith('http')
      ? `${API_BASE_URL}/api/events/${eventId}/playbook-draft-check`
      : `${API_BASE_URL}/events/${eventId}/playbook-draft-check`
    const response = await axios.post(apiUrl, { include_playbook: includePlaybook })
    return response.data
  },

  async updateDisposition(eventId: number, disposition: 'noise' | 'ticket_only' | 'case_required', reason?: string): Promise<{ message: string }> {
    const apiUrl = API_BASE_URL.startsWith('http')
      ? `${API_BASE_URL}/api/events/${eventId}/disposition`
      : `${API_BASE_URL}/events/${eventId}/disposition`
    const response = await axios.post(apiUrl, { disposition, reason })
    return response.data
  }
}

File: backend/api/schemas/events.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from api.schemas.common import PageMeta


class EventRecord(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_event_id: Optional[int] = None
    source: str
    source_label: Optional[str] = None
    source_category: Optional[str] = None
    event_type: Optional[str] = None
    signal_key: Optional[str] = None
    disposition: Optional[str] = None
    disposition_reason: Optional[str] = None
    decision_confidence: Optional[float] = None
    cluster_key: Optional[str] = None
    correlation_key: Optional[str] = None
    signal_family: Optional[str] = None
    external_event_id: Optional[str] = None
    dedup_key: str
    site_id: Optional[int] = None
    netbox_device_id: Optional[int] = None
    host: Optional[str] = None
    name: str
    severity: str
    severity_level: int
    status: str
    acknowledged: bool
    occurred_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    last_seen_at: Optional[datetime] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    case_id: Optional[int] = None
    case_code: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class EventIngestRequest(BaseModel):
    source: str = Field(default="ELK", max_length=50, description="Only ELK or ZABBIX are accepted")
    event_type: str = Field(default="log_signal", max_length=100, description="Only log_signal or zabbix_alert are accepted")
    external_event_id: Optional[str] = Field(default=None, max_length=128)
    site_id: Optional[int] = None
    netbox_device_id: Optional[int] = None
    host: Optional[str] = Field(default=None, max_length=255)
    name: str = Field(..., min_length=1, max_length=512)
    severity: Optional[str] = Field(default=None, max_length=30)
    severity_level: Optional[int] = Field(default=None, ge=0, le=5)
    occurred_at: Optional[datetime] = None
    fingerprint: Optional[str] = Field(default=None, max_length=256)
    tags: List[str] = Field(default_factory=list)
    raw_payload: Dict[str, Any] = Field(default_factory=dict)

class EventIngestResponse(BaseModel):
    accepted: bool = True
    observe_only: bool
    event: EventRecord
    case_id: Optional[int] = None
    case_code: Optional[str] = None


class EventListResponse(BaseModel):
    page: PageMeta
    events: List[EventRecord] = Field(default_factory=list)


class EventClusterItem(BaseModel):
    cluster_key: str
    correlation_key: str
    title: str
    event_count: int
    source_categories: List[str] = Field(default_factory=list)
    dispositions: Dict[str, int] = Field(default_factory=dict)
    case_count: int = 0
    ticket_count: int = 0
    highest_severity: str = "warning"
    host: Optional[str] = None
    site_id: Optional[int] = None
    netbox_device_id: Optional[int] = None
    latest_occurred_at: Optional[datetime] = None
    signal_family: Optional[str] = None
    device_name: Optional[str] = None
    device_role: Optional[str] = None
    site_name: Optional[str] = None
    topology_hint: Optional[str] = None
    root_cause_candidate: Optional[str] = None
    adjacent_devices: List[str] = Field(default_factory=list)
    link_count: int = 0
    impact_scope: Optional[str] = None


class EventClusterListResponse(BaseModel):
    clusters: List[EventClusterItem] = Field(default_factory=list)


class RootCauseCandidateItem(BaseModel):
    candidate_key: str
    title: str
    root_cause_candidate: str
    site_name: Optional[str] = None
    signal_family: Optional[str] = None
    score: float
    ranking_reason: str
    merged_cluster_count: int = 0
    event_count: int = 0
    case_count: int = 0
    ticket_count: int = 0
    source_categories: List[str] = Field(default_factory=list)
    adjacent_devices: List[str] = Field(default_factory=list)
    representative_device: Optional[str] = None
    impact_scope: Optional[str] = None
    recommended_actions: List[Dict[str, Any]] = Field(default_factory=list)


class RootCauseCandidateListResponse(BaseModel):
    items: List[RootCauseCandidateItem] = Field(default_factory=list)


class EventDispatchRequest(BaseModel):
    reviewer: Optional[str] = "system"


class EventDispositionRequest(BaseModel):
    disposition: str = Field(..., description="noise|ticket_only|case_required")
    reason: Optional[str] = None


class EventDispatchResponse(BaseModel):
    success: bool
    message: str
    task_id: Optional[int] = None
    case_id: Optional[int] = None
    case_code: Optional[str] = None
    playbook_check: Dict[str, Any] = Field(default_factory=dict)


class EventTicketCreateRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = "P3"
    requester: Optional[str] = "netops-automation"


class EventTicketResponse(BaseModel):
    success: bool
    message: str
    ticket_id: Optional[str] = None
    provider: Optional[str] = None


class EventTaskLinkItem(BaseModel):
    task_id: int
    task_code: str
    status: str
    source_model: Optional[str] = None
    case_id: Optional[int] = None
    created_at: Optional[datetime] = None


class EventCaseLinkItem(BaseModel):
    case_id: int
    case_code: str
    created_at: Optional[str] = None


class EventRelationsResponse(BaseModel):
    event_id: int
    ticket: Dict[str, Any] = Field(default_factory=dict)
    linked_case: Optional[EventCaseLinkItem] = None
    linked_tasks: List[EventTaskLinkItem] = Field(default_factory=list)


class EventPlaybookDraftRequest(BaseModel):
    include_playbook: bool = True


class EventPlaybookDraftResponse(BaseModel):
    success: bool
    message: str
    event_id: int
    playbook_check: Dict[str, Any] = Field(default_factory=dict)
    playbook_yaml: str = ""


# Backward-compatible alias during event-domain migration.
AlertEventItem = EventRecord
